src/main/java/org/opensearch/geospatial/ip2geo/action/PutDatasourceRequest.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.geospatial.ip2geo.action;
+
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.util.Locale;
+
+import lombok.Getter;
+import lombok.Setter;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.action.ActionRequestValidationException;
+import org.opensearch.action.support.master.AcknowledgedRequest;
+import org.opensearch.common.io.stream.StreamInput;
+import org.opensearch.common.io.stream.StreamOutput;
+import org.opensearch.common.unit.TimeValue;
+import org.opensearch.core.ParseField;
+import org.opensearch.core.xcontent.ObjectParser;
+import org.opensearch.geospatial.ip2geo.common.DatasourceManifest;
+
+/**
+ * GeoIP datasource creation request
+ */
+@Getter
+@Setter
+public class PutDatasourceRequest extends AcknowledgedRequest<PutDatasourceRequest> {
+    private static final Logger LOGGER = LogManager.getLogger(PutDatasourceRequest.class);
+    private static final ParseField ENDPOINT_FIELD = new ParseField("endpoint");
+    private static final ParseField UPDATE_INTERVAL_IN_DAYS_FIELD = new ParseField("update_interval_in_days");
+    /**
+     * @param id Datasource name
+     * @return Datasource name
+     */
+    private String id;
+    /**
+     * @param endpoint URL to a manifest file for a datasource
+     * @return URL to a manifest file for a datasource
+     */
+    private String endpoint;
+    /**
+     * @param updateIntervalInDays Update interval of a datasource
+     * @return Update interval of a datasource
+     */
+    private TimeValue updateIntervalInDays;
+
+    /**
+     * Parser of a datasource
+     */
+    public static final ObjectParser<PutDatasourceRequest, Void> PARSER;
```

Comment:
why this is public?

Reply:
> This is a big PR, is there a way we can split the PR into smaller PRs?

Sorry for the big PR. Those are all new files so splitting will end up with multiple PR with each package mostly. Will try to make PR smaller in the next implementation.

Reply:
It is used in `RestPutDatasourceAction`